cylc/flow/task_pool.py

```diff
@@ -974,76 +1046,66 @@ def check_abort_on_task_fails(self):
-            for output in itask.state.outputs.get_completed():
-                all_task_outputs.add((itask.tdef.name,
-                                      str(itask.point),
-                                      output))
-        for itask in self.get_tasks():
-            # Try to satisfy itask if not already satisfied.
-            if itask.state.prerequisites_are_not_all_satisfied():
-                itask.state.satisfy_me(all_task_outputs)
-
-    def force_spawn(self, itask):
-        """Spawn successor of itask."""
-        if itask.has_spawned:
-            return None
-        itask.has_spawned = True
-        LOG.debug('[%s] -forced spawning', itask)
-        next_point = itask.next_point()
-        if next_point is None:
+        if self.config.initial_point and point < self.config.initial_point:
+            LOG.warning(
+                'Not spawning %s.%s: behind initial cycle point', name, point)
             return
-        new_task = TaskProxy(
-            itask.tdef, start_point=next_point, stop_point=itask.stop_point)
-        return self.add_to_runahead_pool(new_task)
-
-    def spawn_all_tasks(self):
-        """Spawn successors of tasks in pool, if they're ready.
-
-        Return the number of spawned tasks.
-        """
-        n_spawned = 0
-        for itask in self.get_tasks():
-            # A task proxy is never ready to spawn if:
-            #    * it has spawned already
-            #    * its state is submit-failed (avoid running multiple instances
-            #      of a task with bad job submission config).
-            # Otherwise a task proxy is ready to spawn if either:
-            #    * self.tdef.spawn ahead is True (results in spawning out to
-            #      max active cycle points), OR
-            #    * its state is >= submitted (allows successive instances
-            #      to run concurrently, but not out of order).
-            if (
-                not itask.has_spawned
-                and not itask.state(TASK_STATUS_SUBMIT_FAILED)
-                and (
-                    itask.tdef.spawn_ahead
-                    or itask.state(TASK_STATUS_EXPIRED)
-                    or (
-                        not itask.state.is_held
-                        and itask.state.is_gt(TASK_STATUS_READY)
-                    )
-                )
-            ):
-                if self.force_spawn(itask) is not None:
-                    n_spawned += 1
-        return n_spawned
+        elif self.config.final_point and point > self.config.final_point:
+            LOG.warning(
+                'Not spawning %s.%s: beyond final cycle point', name, point)
+            return
+        itask = None
+        for jtask in self.get_all_tasks():
+            if jtask.tdef.name == name and jtask.point == point:
+                itask = jtask
+                # Already spawned.
+                break
+        if itask is None:
+            for tname in self.config.get_task_name_list():
```

Comment:
Yes, that's top of "Future Enhancements" in the proposal: https://cylc.github.io/cylc-admin/proposal-spawn-on-d.html#future-enhancements

> Refactor the task pool (which remains largely as for SoS) to avoid (most) iteration over tasks?

At the moment we have a similar number of loops over the task pool in SoS and SoD. **However** note that on this branch the task pool in large workflows will be relatively much much smaller.